cocotb/share/lib/simulator/simulatormodule.c

```diff
@@ -566,30 +676,30 @@ static PyObject *iterate(PyObject *self, PyObject *args)
 static PyObject *next(PyObject *self, PyObject *args)
 {
     COCOTB_UNUSED(self);
-    gpi_iterator_hdl hdl;
+    gpi_iterator_hdl_Object* hdl_obj;
     gpi_sim_hdl result;
     PyObject *res;
 
-    if (!PyArg_ParseTuple(args, "O&", gpi_iterator_hdl_converter, &hdl)) {
+    if (!PyArg_ParseTuple(args, "O!", &gpi_iterator_hdl_Type, &hdl_obj)) {
         return NULL;
     }
 
-    // It's valid for iterate to return a NULL handle, to make the Python
+    // TODO:eric-wieser: It's valid for iterate to return a NULL handle, to make the Python
     // intuitive we simply raise StopIteration on the first iteration
-    if (!hdl) {
-        PyErr_SetNone(PyExc_StopIteration);
-        return NULL;
-    }
+    // if (!hdl) {
+    //     PyErr_SetNone(PyExc_StopIteration);
+    //     return NULL;
+    // }
```

Comment:
So,
1. `iterate` is not defined on the object and `gpi_iterate` returns `NULL`
2. `gpi_iterator_hdl_New` returns `None` if the iterator is `NULL`,
3. `_SimIterator` sets `._iter` to `None` 
4. On the first iteration, `next` gets `None`
5. `PyArg_ParseTuple` above results in `TypeError`

Seems like you could get rid of this, no `gpi_iterator_hdl_Type` will have a `NULL` handle. But maybe you should add a check to `iterate` to check the outcome of `gpi_iterate` instead?